snapshotter: add more assertion

AkihiroSuda opened the pull request:
This PR adds new assertions to make sure that the content of the parent committed snapshot is propagated to the child active one.

I added `testutil.AssertTree()` for asserting file hiearchy in a declarative way.
I think we should use [continuity](https://github.com/stevvooe/continuity) in the future instead of this utility function.
(requires https://github.com/stevvooe/continuity/issues/48)

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda commented on snapshot/snapshotter.go:
```diff
@@ -282,6 +285,38 @@ func checkSnapshotterBasic(t *testing.T, snapshotter Snapshotter, work string) {
+	aMode, bMode, cMode := os.ModeDir|0755, os.ModeDir|0755, os.ModeDir|0755
+	info1 := &testutil.FileInfo{
+		BaseName: "",
+		// If you set Mode to os.ModeDir | 0777 for "", overlay fails (btrfs and naive passes)
```
I don't know this is expected

dmcgowan replied:
The root is not usually included in these types of comparisons.

As for overlay, my understanding is overlay will overwrite the permission of the mount location with the root of the overlay. Tried changing the permission of the root to match the expectations but still getting a failure.

dmcgowan replied:
Confirmed that doing a `os.Chmod` on the directories post mount gets it to pass. My guess is the root permission is not propagated from unmount -> commit -> prepare -> mount of the new layer because the upper will get recreated with permission 0700, see https://github.com/docker/containerd/blob/master/snapshot/overlay/overlay.go#L192.

dmcgowan commented:
Let me look into this mode issue. It was my intent to use https://github.com/docker/containerd/blob/master/fs/fstest/compare.go for doing these types of comparisons. This comparison function provides a detailed message of what the differences between two directories are using the continuity manifest.

stevvooe commented on testutil/file.go:
```diff
@@ -0,0 +1,117 @@
+
+// FileInfo is a handy struct for asserting information related to file.
+// TODO: use continuity? ( requires https://github.com/stevvooe/continuity/issues/48 )
+type FileInfo struct {
```
There is already an fstest package in this project. Please use that.

stevvooe replied:
@dmcgowan PTAL

dmcgowan replied:
If we change the setup to use the apply methods we can use the method of applying inside the snapshot, then applying a temp directory and comparing the result. I prefer that method to defining the file info like this, makes it much easier to generate more complicated tests without exploding the amount of test definition.

dmcgowan commented:
Not a fan of this method for doing assertions.

```
    --- FAIL: TestOverlay/Basic (0.04s)
    	file.go:115: expected mode drwxrwxrwx, got drwx------ for "/tmp/snapshot-suite-Overlay-926228011/work/nextlayer" (&{BaseName: Mode:drwxrwxrwx Content:[] Target:<nil> Children:[0xc420018c30 0xc420018c80]})
```

I get a lot of contextual information but not the line within `checkSnapshotterBasic` which actually made the call.

AkihiroSuda commented:
thank you for feedback and analysis, updated PR

dmcgowan commented on testutil/fstest/file.go:
```diff
@@ -56,6 +56,13 @@ func CreateDir(name string, perm os.FileMode) Applier {
 }
 
+// RemoveDir returns a file applier which removes the provided directory name
+func RemoveDir(name string) Applier {
```
Wondering if I should have just named `RemoveFile` as `RemoveAll` instead. We shouldn't have functions that are equivalent like this. Let's just rename `RemoveFile` to `Remove` and use that.

stevvooe replied:
Maybe, it should reflect shallow `Remove` and `RemoveAll`. Those are familiar semantics.

dmcgowan commented on testutil/fstest/compare.go:
```diff
@@ -35,3 +38,17 @@ func CheckDirectoryEqual(d1, d2 string) error {
 }
+
+// CheckDirectoryEqualWithApplier compares directory against applier
+func CheckDirectoryEqualWithApplier(a Applier, d string) error {
```
Arguments feel backward given function name

stevvooe replied:
`CompareXXX`?

dmcgowan commented on testutil/fstest/compare.go:
```diff
@@ -35,3 +38,17 @@ func CheckDirectoryEqual(d1, d2 string) error {
+
+// CheckDirectoryEqualWithApplier compares directory against applier
+func CheckDirectoryEqualWithApplier(a Applier, d string) error {
+	// TODO: cache applied
```
Could you explain this more, not sure what the use case for caching the temp directory is.

AkihiroSuda replied:
If this function is called multiple times against identical `Applier`, caching the content of `applied` would be better for optimal performance.

Of course, ideally we should not even create `applied` directory. I'll try to redesign this.

dmcgowan commented on snapshot/snapshotter.go:
```diff
@@ -308,19 +322,14 @@ func checkSnapshotterBasic(t *testing.T, snapshotter Snapshotter, work string) {
 
-	if err := ioutil.WriteFile(filepath.Join(next, "bar"), []byte("bar\n"), 0777); err != nil {
+	if err := fstest.CheckDirectoryEqualWithApplier(initialApplier, next); err != nil {
 		t.Fatal(err)
```
I know this pattern is not used throughout this package but when using the check directory equal function, I find it helpful to fail with a message of what was checked and with `%+v` to pick up as much of the failure trace as possible from `github.com/pkg/errors`

stevvooe commented on fs/copy_test.go:
```diff
@@ -6,7 +6,7 @@ import (
 	_ "crypto/sha256"
 
-	"github.com/docker/containerd/fs/fstest"
+	"github.com/docker/containerd/testutil/fstest"
```
I'm not sure I'm on board with this package name. I really don't like expanding `util` packages. They end up with no clear purpose and just become a dumping ground.

AkihiroSuda replied:
`testutil` itself was newly not added in this PR. Do you mean I should revert `testutil/fstest` to `fs/test`?

dmcgowan replied:
I think keeping it as `fs/fstest` makes sense. `fstest` is the more specific package name, under `fs`. I agree with Stephen's comment that a general testutil package is not well bounded and easily just sucks in more code.

AkihiroSuda commented:
Updated PR

* Introduced `Remove` and `RemoveAll`, which behave similar to `os.Remove` and `os.RemoveAll`
* Now I added `Verify()` to the `Applier` interface. I think this is better abstraction than before (`CheckDirectoryEqualWithApplier`).

AkihiroSuda commented:
Also added `tFatal` which calls `t.Fatal` with stack trace

dmcgowan commented on fs/fstest/file.go:
```diff
@@ -12,6 +12,7 @@ import (
 // Applier applies single file changes
 type Applier interface {
 	Apply(root string) error
+	Verify(root string) error
```
I don't think this interface is correct. An applier does not have the ability to isolate it's change in a way to properly verify a change was applied. This would require passing in both a base directory and a changed directory. However doing this does not belong in the interface and better left to a helper.

For further context, the implemented `Verify` function is only valid for an applier that works on a root, but every incremental Applier has an invalid Verify function.

dmcgowan commented on snapshot/snapshotter.go:
```diff
@@ -253,82 +258,91 @@ func makeTest(t *testing.T, name string, snapshotterFn func(root string) (Snapsh
 
-	if err := os.RemoveAll(filepath.Join(next, "a", "b")); err != nil {
-		t.Log(err)
+	if err := diffApplier.Apply(work); err != nil {
```
Shouldn't this be applied onto `next`? Also would expect a directory check post commit.

AkihiroSuda commented:
@dmcgowan

updated PR, sorry for being late

dmcgowan commented on snapshot/snapshotter.go:
```diff
@@ -256,74 +261,84 @@ func makeTest(t *testing.T, name string, snapshotterFn func(root string) (Snapsh
 }
 
+func tFatal(t *testing.T, err error) {
+	t.Fatal(errors.WithStack(err))
```
`errors.WithStack` is just going to set the stack at this current location, which is not exactly what we want, calling `t.Fatal` directly already gives the failure location. Also using `errors.WithStack` will not cause the error to be formatted with the stack. I think it is cleaner to just use `t.Fatalf("failure reason: %+v", err)` rather calling a helper.

dmcgowan commented:
I think we should remove the `tFatal` function, but everything else LGTM. I would just change `tFatal` back to `t.Fatal` for now and making the messages prettier can be in a separate PR. I think the code where most those failures occur will end up getting refactored anyway so there will not be as much repeating (I plan on porting over functions like these https://github.com/dmcgowan/dsdbench/blob/master/test_util.go#L24).

AkihiroSuda commented:
updated PR @dmcgowan

dmcgowan commented:
LGTM

stevvooe commented:
LGTM

Does this cover any cases enumerated in #511?

AkihiroSuda commented:
Yes marked as 523

stevvooe commented:
@AkihiroSuda The checkboxes. Does this cover any of them?

AkihiroSuda commented:
I meant this check box 😅

> make sure the content of parent appears in the child (#523)
